File: 2nd_quarter/week03/whaeun/광물 캐기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

int solution(vector<int> picks, vector<string> minerals) {
    int answer = 0;
    
    vector<int> mine;
    
    vector<pair<int,int>> mine_sum;
    vector<pair<int,int>> picks_order;
    
    int dia = picks[0];
    int iron = picks[1];
    int stone = picks[2];
    
    int sum = 0;
    
    for(int i = 0; i < minerals.size(); i++){
        if(minerals[i] == "diamond"){
            mine.push_back(3);
        }
        else if(minerals[i] == "iron"){
            mine.push_back(2);
        }
        else{
            mine.push_back(1);
        }
        
        sum += pow(5, mine[i] - 1);
        
        if(i % 5 == 4){
            mine_sum.push_back({sum, i / 5});
            sum = 0;
        }
        
    }
    
    if(sum != 0){
        mine_sum.push_back({sum, minerals.size() / 5});
    }
    
    sort(mine_sum.begin(), mine_sum.end());
    
    for(int i = mine_sum.size() - 1; i >= 0; i--){
        if(dia > 0){
            picks_order.push_back({mine_sum[i].second, 3});
            dia -= 1;
            continue;
        }
        
        if(iron > 0){
            picks_order.push_back({mine_sum[i].second, 2});
            iron -= 1;
            continue;
        }
        
        if(stone > 0){
            picks_order.push_back({mine_sum[i].second, 1});
            stone -= 1;
            continue;
        }
    }
    
    sort(picks_order.begin(), picks_order.end());
    
    int idx = 0;
    
    for(int i = 0; i < min(mine.size(), picks_order.size() * 5); i++){
        if(picks_order[idx].second >= mine[i]){
            answer += 1;
        }
        else{
            answer += pow(5,  mine[i] - picks_order[idx].second);
        }
        
        if(i % 5 == 4){
            idx++;
        }
    }
    
    return answer;
}

File: 2nd_quarter/week03/Royce/광물 캐기.py
def solution(picks, minerals):
    answer = 0
    
    #     다이아 철 돌
    # 다이아 1   1  1
    #  철   5   1  1
    #  돌  25   5  1
    
    # 사용할 수 있는 곡괭이 중 아무거나 하나를 선택
    # 한 번 사용하기 시작한 곡괭이는 사용할 수 없을 때까지 사용
    # 주어진 순서대로만 가능
    
    num = 0
    for k in picks:
        num += k
        
    gok = num * 5
    
    # minerals의 배열의 개수가 num보다 많을 경우 캘 수 있는 만큼만 자른다.
    if len(minerals) > num:
        minerals = minerals[:gok]
    
    next = [[0,0,0] for _ in range((len(minerals) // 5 + 1))]
    # 다음 광물을 위해서 배열을 만들어준다.
    
    for i in range(len(minerals)):
        if minerals[i] == 'diamond':
            next[i // 5][0] += 1
        elif minerals[i] == 'iron':
            next[i // 5][1] += 1
        elif minerals[i] == 'stone':
            next[i // 5][2] += 1
        # 다이아인 경우, 철, 돌인 경우 각각 1씩 더한다.
    
    next.sort(key = lambda x : (-x[0], -x[1], -x[2]))
    # 다이아몬드, 철, 돌 순서로 정렬한다.
    
    for k in next:
        diamond, iron, stone = k
        
        for t in range(len(picks)):
            if picks[t] > 0 and t == 0:
                picks[t] -= 1
                answer += diamond + iron + stone
                break
            elif picks[t] > 0 and t == 1:
                picks[t] -= 1
                answer += (5 * diamond) + iron + stone
                break
            elif picks[t] > 0 and t == 2:
                picks[t] -= 1
                answer += (25 * diamond) + (5 * iron) + stone
                break
                
    
    return answer
